Fix #18897: Acceptance tests for Practice Question Reviewer

imchristie opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #18896.
2. This PR does the following: Implement acceptance tests to test the following:
       5.1 Users can review a question on a topic and Accept it
       5.2 Users can review a question in a topic and Reject it

Currently, It only has the functionality of:
   - assign the right to review questions to the user as the question admin 
   - check if the user does the right to review questions.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] My PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).

## Proof that changes are correct
<img width="721" alt="Proof of Test Passed" src="https://github.com/oppia/oppia/assets/122439444/c1f81ad1-63f8-4b42-9e08-e84c7260b490">

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

imchristie committed:
Merge pull request #1 from oppia/develop

Add new model for Pinning Lesson (#18873)

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
Added test for i18n format

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
fixing the merge

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
Question Admin

imchristie committed:
Added Functions for Question Admin

imchristie committed:
Added Functions for Question Admin

imchristie committed:
Added Functions for Question Admin

imchristie committed:
Added Functions for Question Admin

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
Update develop

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
Adding Question Admin Functions

DubeySandeep commented:
@U8NWXD Can you take a pass on this PR? (I'm  looking at other PRs and doesn't have much time today!)

U8NWXD commented on core/tests/puppeteer-acceptance-tests/puppeteer-testing-utilities/puppeteer-utils.js:
```diff
@@ -47,7 +47,7 @@ module.exports = class baseUser {
          * headless mode. As per the expected behavior we need to make sure
          * every test passes on both modes. */
-        headless: false,
+        headless: true,
```
Does this mean #17761 will be fixed by this PR?

imchristie replied:
No, The test doesn't work if headless is false so I have to change it to true. #17761 was requesting to fix it so tests can pass in true and false.

seanlip replied:
It seems that other existing acceptance tests work with headless=false -- at least, they pass on CI. Why doesn't this test work with headless=false? (Asking because the file you're editing here is a utility file that affects all acceptance tests, right?)

U8NWXD replied:
I agree with @seanlip. This change will affect all tests, so we need to pick one mode (headless=true or headless=false) and make sure that all the tests pass in that mode

imchristie replied:
On my pc, if headless=false, an error saying "Missing X server or $DISPLAY" will pop up therefore I changed it to true. But since other existing acceptance tests work with headless=false then I'm guessing it's a problem in the setting on my pc. I will go ahead and fix the comment and push the commit with headless=false and see if it works.

imchristie replied:
@U8NWXD @seanlip I have pushed a new commit with headless=false and the tests are working fine now. All checks passed.

imchristie replied:
Done

U8NWXD commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,163 @@
+  '../puppeteer-testing-utilities/puppeteer-utils.js');
+const testConstants = require(
+  '../puppeteer-testing-utilities/test-constants.js');
+const ContributorDashboardAdminUrl =
```
Please make this variable camelCase since it's not a class

imchristie replied:
Thanks for pointing it out, It's fixed now.

U8NWXD commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,163 @@
+    );
+    if (displayedUsers.includes(username)) {
+      throw new Error(
+        `${username} still has the right to review question!`);
```
This function could also be used as a pre-check (i.e. to ensure that at the start of the test, the user doesn't have reviewing rights), but in that scenario, the word "still" wouldn't make sense. To accommodate this use case, consider re-wording the message to `${username} has the right to review questions!`

imchristie replied:
Thanks for pointing it out, It's fixed now.

U8NWXD commented on core/tests/puppeteer-acceptance-tests/spec/practice-question-reviewer-tests/review-and-accept-question.spec.js:
```diff
@@ -0,0 +1,54 @@
+      await userFactory.closeBrowserForUser(reviewer);
+      await questionAdmin.navigateToContributorDashboardAdminPage();
+
+      await questionAdmin.addReviewQuestionRights('reviewer');
```
Please add pre-checks to ensure that the user initially has neither reviewing nor contribution rights

imchristie replied:
Thanks for pointing it out, I added a line to call the function expectUserToNotBeDisplayed now.

U8NWXD replied:
I think `expectUserToNotBeDisplayed()` requires that `viewContributorReviewQuestionRightsByRole()` be called first, right? Otherwise the user won't be displayed simply because we're not on the page that displays privileged users.

More generally, I think it's confusing to separate the functions that navigate to the page that displays user rights from the functions that check those rights. This separation makes it easy to forget to call the navigation function. Instead, would it be clearer to have a single function that both navigates to the appropriate page and checks permissions?

seanlip replied:
To add to this -- I think the comment in https://github.com/oppia/oppia/pull/19178#discussion_r1409594832 might help address this issue. 

(Also, I agree with @U8NWXD's point generally -- @agallop @imchristie @StephenYu2018 this is probably something to bear in mind when writing top-level tests.)

imchristie replied:
Yes @U8NWXD, you are right. viewContributorReviewQuestionRightsByRole() should be called after expectUserToNotDisplayed() and I agree that it might be confusing to separate it. I will combine them together like @seanlip's comment.

imchristie replied:
Done

U8NWXD commented:
Sorry for the delay in reviewing @imchristie!

imchristie commented:
@U8NWXD Thank you for the review and suggestions and I will work on them very soon. This PR is not intended to pass and be merged as there are still some missing functionalities. It is intended to show my process to the acceptance tests team.

seanlip commented:
Hi @imchristie -- this PR actually needs to get merged. It is fine if we do the relevant feature in stages but you should get this PR fully functioning so that we can merge it into develop. Please fix the comments as needed and ensure the necessary checks are passing, thanks.

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
Merge branch 'develop' into christie_pqr

imchristie committed:
All comments fixed

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/practice-question-reviewer-tests/review-and-accept-question.spec.js:
```diff
@@ -0,0 +1,55 @@
+      let reviewer = await userFactory.createNewGuestUser(
+        'reviewer', 'review@example.com');
+      await userFactory.closeBrowserForUser(reviewer);
+      await questionAdmin.navigateToContributorDashboardAdminPage();
```
Let's modify this whole test (and the associated utilities) so that the top-level reads as follows:

```
      await questionAdmin.navigateToContributorDashboardAdminPage();
      await questionAdmin.verifyQuestionReviewersExcludeUser('reviewer');   // this checks the list of question reviewers

      await questionAdmin.addReviewQuestionRights('reviewer');

      await questionAdmin.verifyUserCanReviewQuestions('reviewer');   // this checks the user's roles
      await questionAdmin.verifyQuestionReviewersIncludeUser('reviewer');  // this checks the list of question reviewers
```

imchristie replied:
@seanlip, This is done and new functions are made for checking if the user has the right to submit questions or not. Their names follow the structure of the names you suggested.

imchristie replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,163 @@
+   * Function to display contribution rights by user.
+   * @param {string} username - the username of the user to view.
+   */
+  async viewContributionRightsForUser(username) {
```
This function "does nothing", I don't think it should be exposed. If anything it might be a private function that other functions in this file make use of.

imchristie replied:
This function has been removed as it 'does nothing' without calling other functions. I have combined this function and other functions to create functions with clearer purposes.

imchristie replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,163 @@
+   * Function to check if the user is displayed as a question reviewer.
+   * @param {string} username - the user expected to be displayed.
+   */
+  async expectUserToBeDisplayed(username) {
```
This name is vague because it says nothing about practice questions, so it's unclear where we're expecting the user to be displayed. See my comment in the top-level test for alternative naming.

imchristie replied:
This function has been modified like your comment with better naming.

imchristie replied:
Done

seanlip reviewed (changes requested):
Hi @imchristie, thanks, this looks pretty good! I took a pass and have some suggestions, could you PTAL? Once those are addressed, let's try and get this in so that the other team members can build on it.

Also @U8NWXD  could you PTAL and review again? Thanks.

U8NWXD commented:
> @U8NWXD Thank you for the review and suggestions and I will work on them very soon. This PR is not intended to pass and be merged as there are still some missing functionalities. It is intended to show my process to the acceptance tests team.

@imchristie sure, no problem. For the future, if you don't want reviewers to take a look at a PR yet because you're still working on it, you can mark the PR as a draft before you open it so that Oppiabot doesn't assign reviewers

U8NWXD commented:
I left follow-up comments

imchristie committed:
Comments fixed and added all tests needed

imchristie committed:
Comments fixed and added all tests needed

imchristie committed:
Comments fixed and added all tests needed

imchristie committed:
Comments fixed and added all tests needed

imchristie committed:
Comments fixed and added all tests needed

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
Merge branch 'develop' into christie_pqr

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
All comments fixed and files are combined with private functions

imchristie committed:
All comments fixed and files are combined with private functions

imchristie committed:
All comments fixed and files are combined with private functions

imchristie commented:
@seanlip @U8NWXD PTAL

imchristie committed:
Added working private functions

imchristie committed:
Added working private functions

seanlip commented:
@imchristie Please make sure you have responded to all comments from reviewers before requesting a follow-up review. See step 5 of our PR guidelines here: https://github.com/oppia/oppia/wiki/Make-a-pull-request

I am reassigning this PR to you so that you can do this. Thanks!

yashwardhan-jyani committed:
Fix #18192: Removing any extra space from new state names (#19201)

removes extra space on new state

Co-authored-by: Harshvardhan Singh <73544247+Lawful2002@users.noreply.github.com>

imchristie committed:
error still occur

imchristie committed:
error still occur

imchristie commented:
@seanlip, I have responded to all comments now and could you PTAL?

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

seanlip commented:
Hi @imchristie, if you look at the comments on the current page, there are still no replies: https://github.com/oppia/oppia/pull/19178#pullrequestreview-1755667334 . Note that you still need to reply "Done" to comments that you have addressed.

Are your replies marked as "Pending"? If so you will need to publish them using the button at the top-right corner of the Files Changed tab. Could you please check that? (EDIT: Please also do this in the future when you submit requests for review -- take a look at the "Files Changed" page and ensure the changes are what you want them to be, and take a look at the main "Conversation" page for the PR on GitHub and check that all comments have been responded to from your side.

imchristie committed:
Change private functions to public

imchristie committed:
Change private functions to public

imchristie committed:
Merge remote-tracking branch 'upstream/develop' into develop

imchristie committed:
Merge branch 'develop' into christie_pqr

DhairyaMajmudar committed:
Fix #19196 : Rearranging the Skill Prerequisite Modal  (#19204)

* Fixes #19040, Changing Twitter Logo

* lint checks

* bug-responsive-view

* Adding Responsiveness

* Removing the bug in the Android Page

* Rearranging

* Fixing Lint Errors

* Failing Lint Checks

* Failing Lint Checks

* Lint Checks

AFZL210 committed:
Fix #17711: Add Image Upload Prerequisites (#19194)

* Fix #17711: Add Image Upload Prerequisites

* Clean up image upload prerequisite text

AFZL210 committed:
Fix #17951: Align creator dashboard stats elements properly (#19208)

Fix alignment of creator dashboard stats element

imchristie reviewed (commented):
This commit has all the functionalities needed for #18897 and all changes requested are made.

Commit:
Update versions to 3.3.3 (#19253)

* Update core/feconf.py

* Update package.json

---------

Co-authored-by: github-actions[bot] <github-actions@noreply.com>

imchristie commented:
> Hi @imchristie, thanks, this looks pretty good! I took a pass and have some suggestions, could you PTAL? Once those are addressed, let's try and get this in so that the other team members can build on it.
> 
> Also @U8NWXD could you PTAL and review again? Thanks.

All comments are addressed and fixed.

imchristie reviewed (commented):
All comments addressed.

nikitaevg committed:
fix #17759: use skill id from the new exploration (#19247)

* fix #17759: use skill id from the new exploration

* remove fake async

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/practice-question-admin-tests/remove-contribution-rights.spec.js:
```diff
@@ -0,0 +1,60 @@
+      await userFactory.createNewQuestionAdmin('questionAdm');
+  }, DEFAULT_SPEC_TIMEOUT);
+
+  it('should be able to provide rights to review and submit questions to user.',
```
I think you might as well combine this with the test in the previous file (and change its name to add-and-remove-contribution-rights) so that we don't have to bear the setup cost twice, especially since this can quite easily follow on from the previous test. What do you think?

imchristie replied:
@seanlip. I prefer to combine as I think the code will look cleaner if combined. This user story is separated into 6.1 and 6.2 so I split it into two files but I think it's better to combine these two files.

seanlip replied:
OK, please go ahead and combine them. (Also, in future, please do apply editorial discretion to the journeys given in the spreadsheet -- if you think things would be better combined or separated, please ask on the groupchat and we can give you feedback on that. Thanks!)

imchristie replied:
Done. The files are combined.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,317 @@
+  /**
+   * Function to display reviewing questions rights by role
+   * and check if the user is not displayed as a question reviewer
+   * @param {string} usename - the user expected to not be displayed.
```
"username" is misspelled, here and elsewhere.

imchristie replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,317 @@
+    await this.page.waitForNetworkIdle();
+
+    await this.page.waitForSelector(viewContributorSubmitQuestionResult);
+    const displayedUsername = await this.page.$eval(
```
This should be plural since you're retrieving a list and calling "includes" on it, right?

Also, something seems off about the naming here. Why would a list of "displayed usernames"  include the text "Not-allowed"? Can you use a variable name that describes what the list actually contains?

Ditto elsewhere.

imchristie replied:
@seanlip, this function checks if the specific user is allowed to submit questions. The website should show a table like the following:
<img width="610" alt="image" src="https://github.com/oppia/oppia/assets/122439444/f6a948ce-2918-4bf6-bfa4-cef66c2941d9">
The variable displayedUsername should store the value of the result (under "Submit Questions"), which should be either "allowed" or "not-allowed", depends on whether the user is allowed to submit questions. I will change the variable name to displayResult as it is confusing.

seanlip replied:
Thanks -- yup the current name is confusing and doesn't describe the div properly. But displayResult is pretty vague. I suggest that you extract the value from the list and call it something like questionReviewStatusForUser or similar.

imchristie replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,317 @@
+    const displayedUsers = await this.page.$eval(
+      viewRoleUserResult,
+      element => element.innerText
+    );
```
I think it would be a good idea to have an async private function that returns the list of displayed users for the submit case, and another that does the same for the review case -- then these public functions can use the results from those private ones.  That way you avoid duplicating large chunks of code, and the returned list of users seems like a sensible common point to factor the logic out.

imchristie replied:
@seanlip, It is a good idea. I will go ahead of make the new functions. What should the names of the functions be? Will viewQuestionSubmitters and viewQuestionReviewers do?

seanlip replied:
Usually for functions that return something I would start them with `get`. So perhaps getDisplayedListOfQuestionSubmitters() or similar -- thoughts?

That way, it's clear it's actually returning something and not just a passive "viewing" action.

imchristie replied:
Done

imchristie replied:
I created two functions that return the list of displayed users with rights of question reviewing and submitting, and also a new function called contributionStatusForUser that takes the username as an argument and returns the status of the contribution rights of the user.

seanlip reviewed (changes requested):
Thanks @imchristie, took another pass. PTAL.

seanlip commented:
Also @imchristie just a note, please make sure the list of Assignees is correct when you want a re-review (it should be updated with the list of reviewer usernames, and you should remove yourself when you have no further immediate actions to take on the PR). Thanks!

Also @U8NWXD PTAL as codeowner. Thanks!

xantric committed:
Fix #14937  : Removed python-matplotlib from scripts/install_prerequisites.sh (#19244)

removed python-matplotlib from scripts/install_prerequisites.sh as python-matplotlib does not have an installation candidate

imchristie committed:
Update question-admin-utils.js

Fixed the typo of "usename" to "username"

U8NWXD commented on core/tests/puppeteer-acceptance-tests/spec/practice-question-admin-tests/remove-contribution-rights.spec.js:
```diff
@@ -0,0 +1,60 @@
+        'contributor', 'contributor@example.com');
+      await userFactory.closeBrowserForUser(contributor);
+
+      await questionAdmin.navigateToContributorDashboardAdminPage();
```
This test also needs a pre-check

imchristie replied:
@U8NWXD  Thank you for the suggestion, I will add a pre-check if the files will not be combined.

imchristie replied:
The files are combined and a pre-check is added before the rights are added to the user.

seanlip replied:
@imchristie I don't think you have pushed your changes -- the files still look separate to me. In the future, please could you check the "Files Changed" tab and verify that your changes are correct before requesting a follow-up review? Thanks.

imchristie replied:
@seanlip I was trying to push my changes but a linter error showed up. 
<img width="494" alt="image" src="https://github.com/oppia/oppia/assets/122439444/279fe841-3082-47f8-b886-b4428bc32e9e">
This error never occurred before and now it stops me from pushing my changes. I ran the same linter check on other existing files (super-admin-utils.js, translation-admin-utils.js, blog-post-admin-utils.js) and this error also occurred. The "element.innerText" is not a file but the linter check somehow counts it as a file name.

imchristie replied:
Here is a more detailed screenshot to explain the error.
<img width="500" alt="image" src="https://github.com/oppia/oppia/assets/122439444/afc8ef23-67df-41fd-b3a7-3da1b6f8b9ca">

U8NWXD reviewed (dismissed):
Almost there! I just have one more comment aside from the notes seanlip already left.

merryyshen committed:
Fix #16559: Added sort for mis-ordering of exploration states again (#19254)

Added sort for mis-ordering of exploration states #19246

imchristie reviewed (commented):
All comments addressed and waiting for suggestions from @seanlip.

seanlip commented:
Also @imchristie a further reminder that when requesting a review, please update the Assignees field so that the reviewers are assigned (and, if you don't have further actions, that you are not assigned). Thanks.

imchristie commented:
> Also @imchristie a further reminder that when requesting a review, please update the Assignees field so that the reviewers are assigned (and, if you don't have further actions, that you are not assigned). Thanks.

I tried to push my changes but a linter error occurred, which stopped me from pushing my changes. I believe it is flaky as the error should not occur. After the error is fixed, I will push the changes, unassign myself, and assign reviewers.

seanlip commented:
@imchristie Do you by any chance have a file called element.innerText in your Oppia directory?

imchristie commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,317 @@
+    await this.page.waitForSelector(viewRoleUserResult);
+    const displayedUsers = await this.page.$eval(
+      viewRoleUserResult,
+      element => element.innerText
```
> Do you by any chance have a file called element.innerText in your Oppia directory?
> […](#)
> On Mon, Dec 11, 2023, 12:07 imchristie ***@***.***> wrote: Also @imchristie <https://github.com/imchristie> a further reminder that when requesting a review, please update the Assignees field so that the reviewers are assigned (and, if you don't have further actions, that you are not assigned). Thanks. I tried to push my changes but a linter error occurred, which stopped me from pushing my changes. I believe it is flaky as the error should not occur. After the error is fixed, I will push the changes, unassign myself, and assign reviewers. — Reply to this email directly, view it on GitHub <[#19178 (comment)](https://github.com/oppia/oppia/pull/19178#issuecomment-1849292130)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACQV5SWQ6RJBUKF2QBV743LYI2BHXAVCNFSM6AAAAAA7SOS4LWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNBZGI4TEMJTGA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

@seanlip I don't have a file called element.innerText in my directory. This file and all other files in the user-utilities file mentioned element.innerText to address the text content of the element.

imchristie replied:
This error had been fixed.

imchristie reviewed (commented):
add-contribution-rights.spec.js and remove-contribution-rights.spec.js are merged as add-and-remove-contribution-rights.spec.js. Three new functions are created in question-admin-utils.js to avoid duplicating large chunks of code.

imchristie commented:
@seanlip @U8NWXD PTAL. I believe the failed check is flaky because my code has nothing to do with the additionalEditorFeaturesModals.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,311 @@
+   * Function to display the contribution rights status for the user.
+   * @param {string} username - the username of the user.
+   */
+  async contributionStatusForUser(username) {
```
Please start all function names with verbs. 

Perhaps this one could also extract and return the user's contribution status.

imchristie replied:
Done. This function is also extracted and now it returns an object/element that contains the user's contribution status.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,311 @@
+  /**
+   * Function to display the list of question reviewers
+   */
+  async getDisplayedListOfQuestionReviewers() {
```
Can you actually have these functions return the details of the thing that it's getting? Generally GET functions return something -- in this case it should probably be the relevant list. (As a result, you should be able to move some logic from the caller into here.)

imchristie replied:
Done. These functions return an object/element that contains the list of the users displayed.

seanlip reviewed (changes requested):
Thanks @imchristie! Looks generally good but I have a suggestion for tightening the naming, PTAL.

imchristie committed:
updated functions to return object

imchristie committed:
set headless as false

imchristie reviewed (commented):
All comments addressed

imchristie commented:
@seanlip PTAL

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,308 @@
+
+  /**
+   * Function to return the list of question reviewers
+   * @returns {object} displayedUsers - an element contains the list
```
@imchristie I'm confused about this. What's the type of the return value? Is it a list of strings (rather than "an element containing the list" or some sort of object)?

If it's not a simple list of strings ... why not?

imchristie replied:
The function $eval() returns an element which is wrapped into an ElementHandle so I put "object" but it is also an array with strings. It's a bit confusing so I will change the comments to string for a clearer description of what it returns. 
Also, I changed the comment of the return type of getContributionStatusForUser as it is a string of the result. And I changed the way of changing the return value to make it clearer.

seanlip replied:
Hm, I still don't see any changes in the docstring?

imchristie replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,308 @@
+const roleMethodValue = 'role';
+
+// "View Contributor Dashboard Users" form elements.
+const viewContributorFilterMethodSelect =
```
Maybe this and the other three "select"s below could be called "selector" rather than "select" (since "selector" is more obviously a noun).

imchristie replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,308 @@
+
+    if (displayedUsers.includes(username)) {
+      throw new Error(
+        `${username} has the right to review question!`);
```
question --> questions

imchristie replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/question-admin-utils.js:
```diff
@@ -0,0 +1,308 @@
+
+    if (displayedUsers.includes(username)) {
+      throw new Error(
+        `${username} has the right to submit question!`);
```
question --> questions

imchristie replied:
Done

seanlip commented:
(Sorry @imchristie, missed some comments yesterday. I don't have anything else besides this last round, though!)

imchristie committed:
All comments addressed and changed the way of checking the result

imchristie committed:
All comments addressed and changed the way of checking the result

imchristie committed:
select changed to selector

imchristie committed:
select changed to selector

imchristie reviewed (commented):
All comments addressed.

imchristie commented:
> (Sorry @imchristie, missed some comments yesterday. I don't have anything else besides this last round, though!)

No worries. All comments are addressed now. @seanlip PTAL.

seanlip reviewed (approved):
Looks great, thanks @imchristie! Let's merge this :)
